# Mixed Integer Programming

The following sections present an example of a MIP problem and show how to solve it. Here's the problem:

__Maximize $x + 10y$__ subject to the following constraints:

$x + 7 y	≤	17.5$

$x	≤	3.5$

$x	≥	0$

$y	≥	0$

$x, y$ are integers

Since the constraints are linear, this is just a linear optimization problem in which the solutions are required to be integers. The graph below shows the integer points in the feasible region for the problem.


<img src="https://developers.google.com/optimization/images/mip/feasible_region.png" width="700">

Notice that this problem is very similar to the linear optimization problem, but in this case we require the solutions to be integers.

## Basic steps for solving a MIP problem

To solve a MIP problem, your program should include the following steps :

1. Import the linear solver wrapper,
2. declare the MIP solver,
3. define the variables,
4. define the constraints,
5. define the objective,
6. call the MIP solver and
7. display the solution

__Note:__ If or-tools is not installed on your machine, first install it (Google Colab users always need to install or-tools)

In [ ]:
pip install --upgrade --user ortools

#### 1. Import the linear solver wrapper
Import (or include) the OR-Tools linear solver wrapper, an interface for MIP solvers and linear solvers, as shown below.

In [7]:
from ortools.linear_solver import pywraplp


#### 2. Declare the MIP solver
The following code declares the MIP solver for the problem. This example uses the third-party solver SCIP

In [8]:
# Create the mip solver with the SCIP backend.
solver = pywraplp.Solver.CreateSolver('SCIP')


#### 3. Define the variables
The following code defines the variables in the problem.

In [9]:
infinity = solver.infinity()
# x and y are integer non-negative variables.
x = solver.IntVar(0.0, infinity, 'x')
y = solver.IntVar(0.0, infinity, 'y')

print('Number of variables =', solver.NumVariables())

Number of variables = 2


#### 4. Define the constraints
The following code defines the constraints for the problem.

In [10]:
# x + 7 * y <= 17.5.
solver.Add(x + 7 * y <= 17.5)

# x <= 3.5.
solver.Add(x <= 3.5)

print('Number of constraints =', solver.NumConstraints())

Number of constraints = 2


#### 5. Define the objective
The following code defines the objective function for the problem.

In [11]:
# Maximize x + 10 * y.
solver.Maximize(x + 10 * y)

#### 6. Call the solver
The following code calls the solver.

In [12]:
status = solver.Solve()

#### 7. Display the solution
The following code displays the solution.

In [13]:
if status == pywraplp.Solver.OPTIMAL:
    print('Solution:')
    print('Objective value =', solver.Objective().Value())
    print('x =', x.solution_value())
    print('y =', y.solution_value())
else:
    print('The problem does not have an optimal solution.')

Solution:
Objective value = 23.0
x = 3.0
y = 2.0


__The optimal value of the objective function is 23, which occurs at the point x = 3, y = 2.__

## Complete programs
Here are the complete programs.

In [14]:
from ortools.linear_solver import pywraplp


def main():
    # Create the mip solver with the SCIP backend.
    solver = pywraplp.Solver.CreateSolver('SCIP')

    infinity = solver.infinity()
    # x and y are integer non-negative variables.
    x = solver.IntVar(0.0, infinity, 'x')
    y = solver.IntVar(0.0, infinity, 'y')

    print('Number of variables =', solver.NumVariables())

    # x + 7 * y <= 17.5.
    solver.Add(x + 7 * y <= 17.5)

    # x <= 3.5.
    solver.Add(x <= 3.5)

    print('Number of constraints =', solver.NumConstraints())

    # Maximize x + 10 * y.
    solver.Maximize(x + 10 * y)

    status = solver.Solve()

    if status == pywraplp.Solver.OPTIMAL:
        print('Solution:')
        print('Objective value =', solver.Objective().Value())
        print('x =', x.solution_value())
        print('y =', y.solution_value())
    else:
        print('The problem does not have an optimal solution.')

    print('\nAdvanced usage:')
    print('Problem solved in %f milliseconds' % solver.wall_time())
    print('Problem solved in %d iterations' % solver.iterations())
    print('Problem solved in %d branch-and-bound nodes' % solver.nodes())


if __name__ == '__main__':
    main()

Number of variables = 2
Number of constraints = 2
Solution:
Objective value = 23.0
x = 3.0
y = 2.0

Advanced usage:
Problem solved in 4.000000 milliseconds
Problem solved in 0 iterations
Problem solved in 1 branch-and-bound nodes


## Comparing Linear and Integer Optimization
Let's compare the solution to the integer optimization problem, shown above, with the solution to the corresponding linear optimization problem, in which integer constraints are removed. You might guess that the solution to the integer problem would be the integer point in the feasible region closest to the linear solution — namely, the point $x = 0,  y = 2$. But as you will see next, this is not the case.

You can easily modify the program in the preceding section to solve the linear problem by making the following changes:

In [19]:
from ortools.linear_solver import pywraplp


def main():
    # Create the mip solver with the SCIP backend.
    solver = pywraplp.Solver.CreateSolver('SCIP')

    infinity = solver.infinity()
    # x and y are integer non-negative variables.
    x = solver.NumVar(0.0, infinity, 'x')
    y = solver.NumVar(0.0, infinity, 'y')

    print('Number of variables =', solver.NumVariables())

    # x + 7 * y <= 17.5.
    solver.Add(x + 7 * y <= 17.5)

    # x <= 3.5.
    solver.Add(x <= 3.5)

    print('Number of constraints =', solver.NumConstraints())

    # Maximize x + 10 * y.
    solver.Maximize(x + 10 * y)

    status = solver.Solve()

    if status == pywraplp.Solver.OPTIMAL:
        print('Solution:')
        print('Objective value =', solver.Objective().Value())
        print('x =', x.solution_value())
        print('y =', y.solution_value())
    else:
        print('The problem does not have an optimal solution.')

    print('\nAdvanced usage:')
    print('Problem solved in %f milliseconds' % solver.wall_time())
    print('Problem solved in %d iterations' % solver.iterations())
    print('Problem solved in %d branch-and-bound nodes' % solver.nodes())


if __name__ == '__main__':
    main()

Number of variables = 2
Number of constraints = 2
Solution:
Objective value = 25.0
x = 0.0
y = 2.5

Advanced usage:
Problem solved in 4.000000 milliseconds
Problem solved in 0 iterations
Problem solved in 0 branch-and-bound nodes


The solution to the linear problem occurs at the point x = 0,  y = 2.5, where the objective function equals 25. Here's a graph showing the solutions to both the linear and integer problems.
<img src="https://developers.google.com/optimization/images/mip/feasible_region_sol.png" width="700">

Notice that the integer solution is not close to the linear solution, compared with most other integer points in the feasible region. In general, the solutions to a linear optimization problem and the corresponding integer optimization problems can be far apart. Because of this, the two types of problems require different methods for their solution.